File: src/jvm/package/src/doc-examples/java/questpdf/docexamples/codepatterns/CodePatternDynamicComponentExample.java
package questpdf.docexamples.codepatterns;

import com.questpdf.elements.DynamicContext;
import com.questpdf.elements.IDynamicElement;
import com.questpdf.fluent.Document;
import com.questpdf.helpers.Colors;
import com.questpdf.helpers.PageSizes;
import com.questpdf.helpers.Placeholders;
import com.questpdf.infrastructure.DynamicComponentComposeResult;
import com.questpdf.infrastructure.IDynamicComponent;
import com.questpdf.infrastructure.Size;
import com.questpdf.infrastructure.TextStyle;
import org.junit.jupiter.api.Test;
import questpdf.docexamples.DocExample;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ThreadLocalRandom;
import java.util.stream.Stream;

public class CodePatternDynamicComponentExample extends DocExample {

    @Test
    public void dynamic() {
        var items = Stream.generate(OrderItem::new).limit(25).toList();

        Document
            .create(document -> {
                document.page(page -> {
                    page.size(PageSizes.getA4());
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(50f);

                    page.content()
                        .decoration(decoration -> {
                            decoration.before()
                                .paddingBottom(10f)
                                .text(text -> {
                                    text.defaultTextStyle(TextStyle.getDefault().bold().fontColor(Colors.Blue.getDarken2()));
                                    text.span("Page ");
                                    text.currentPageNumber();
                                    text.span(" of ");
                                    text.totalPages();
                                });

                            decoration.content()
                                .dynamic(new OrdersTableWithPageSubtotalsComponent(items));
                        });
                });
            })
            .generatePdf(output("code-pattern-dynamic-component-table-with-per-page-subtotals.pdf"));
    }

    public static class OrderItem {
        public final String itemName = Placeholders.label();

        // The Java counterpart of C# Placeholders.Random.Next(minValue, maxValue)
        // (System.Random is not bridged): a random integer within [minValue, maxValue).
        public final int price = ThreadLocalRandom.current().nextInt(1, 11) * 10;
        public final int count = ThreadLocalRandom.current().nextInt(1, 11);
    }

    public static class OrdersTableWithPageSubtotalsComponentState {
        public int shownItemsCount = 0;
    }

    public static class OrdersTableWithPageSubtotalsComponent implements IDynamicComponent {

        // The Java counterpart of the Kotlin Pair<OrderItem, IDynamicElement>.
        private record OrderItemRow(OrderItem item, IDynamicElement element) {}

        private final List<OrderItem> items;
        public OrdersTableWithPageSubtotalsComponentState state;

        public OrdersTableWithPageSubtotalsComponent(List<OrderItem> items) {
            this.items = items;

            this.state = new OrdersTableWithPageSubtotalsComponentState();
            this.state.shownItemsCount = 0;
        }

        @Override
        public DynamicComponentComposeResult compose(DynamicContext context) {
            var header = composeHeader(context);
            var sampleFooter = composeFooter(context, List.of());
            var decorationHeight = header.getSize().getHeight() + sampleFooter.getSize().getHeight();

            var rows = getItemsForPage(context, decorationHeight);
            var footer = composeFooter(context, rows.stream().map(OrderItemRow::item).toList());

            var content = context.createElement(container -> {
                container.shrink().decoration(decoration -> {
                    decoration.before().element(header);

                    decoration.content().column(column -> {
                        for (var row : rows)
                            column.item().element(row.element());
                    });

                    decoration.after().element(footer);
                });
            });

            var state = new OrdersTableWithPageSubtotalsComponentState();
            state.shownItemsCount = this.state.shownItemsCount + rows.size();
            this.state = state;

            var result = new DynamicComponentComposeResult();
            result.setContent(content);
            result.setHasMoreContent(this.state.shownItemsCount < items.size());
            return result;
        }

        private IDynamicElement composeHeader(DynamicContext context) {
            return context.createElement(element -> {
                element
                    .width(context.getAvailableSize().getWidth())
                    .borderBottom(1f)
                    .borderColor(Colors.Grey.getDarken2())
                    .padding(10f)
                    .defaultTextStyle(TextStyle.getDefault().semiBold())
                    .row(row -> {
                        row.constantItem(50f).text("#");
                        row.relativeItem().text("Item name");
                        row.constantItem(75f).alignRight().text("Count");
                        row.constantItem(75f).alignRight().text("Price");
                        row.constantItem(75f).alignRight().text("Total");
                    });
            });
        }

        private IDynamicElement composeFooter(DynamicContext context, List<OrderItem> items) {
            var total = items.stream().mapToInt(x -> x.count * x.price).sum();

            return context.createElement(element -> {
                element
                    .width(context.getAvailableSize().getWidth())
                    .padding(10f)
                    .alignRight()
                    .text("Subtotal: " + total + "$")
                    .bold();
            });
        }

        private List<OrderItemRow> getItemsForPage(DynamicContext context, float decorationHeight) {
            var rows = new ArrayList<OrderItemRow>();
            var totalHeight = decorationHeight;

            for (var index = state.shownItemsCount; index < items.size(); index++) {
                var item = items.get(index);
                var itemNumber = index + 1;

                var element = context.createElement(content -> {
                    content
                        .width(context.getAvailableSize().getWidth())
                        .borderBottom(1f)
                        .borderColor(Colors.Grey.getLighten2())
                        .padding(10f)
                        .row(row -> {
                            row.constantItem(50f).text(String.valueOf(itemNumber));
                            row.relativeItem().text(item.itemName);
                            row.constantItem(75f).alignRight().text(String.valueOf(item.count));
                            row.constantItem(75f).alignRight().text(item.price + "$");
                            row.constantItem(75f).alignRight().text(item.count * item.price + "$");
                        });
                });

                var elementHeight = element.getSize().getHeight();

                // it is important to use the Size.Epsilon constant to avoid floating point comparison issues
                if (totalHeight + elementHeight > context.getAvailableSize().getHeight() + Size.Epsilon)
                    break;

                totalHeight += elementHeight;
                rows.add(new OrderItemRow(item, element));
            }

            return rows;
        }
    }
}

File: src/jvm/package/src/doc-examples/kotlin/questpdf/docexamples/codepatterns/CodePatternDynamicComponentExample.kt
package questpdf.docexamples.codepatterns

import com.questpdf.elements.DynamicContext
import com.questpdf.elements.IDynamicElement
import com.questpdf.fluent.Document
import com.questpdf.helpers.Colors
import com.questpdf.helpers.PageSizes
import com.questpdf.helpers.Placeholders
import com.questpdf.infrastructure.DynamicComponentComposeResult
import com.questpdf.infrastructure.IDynamicComponent
import com.questpdf.infrastructure.Size
import com.questpdf.infrastructure.TextStyle
import org.junit.jupiter.api.Test
import questpdf.docexamples.DocExample
import kotlin.random.Random

class CodePatternDynamicComponentExample : DocExample() {

    @Test
    fun dynamic() {
        val items = (0 until 25).map { OrderItem() }

        Document
            .create {
                page {
                    size(PageSizes.A4)
                    defaultTextStyle { fontSize(20f) }
                    margin(50f)

                    content()
                        .decoration {
                            before()
                                .paddingBottom(10f)
                                .text {
                                    defaultTextStyle(TextStyle.Default.bold().fontColor(Colors.Blue.Darken2))
                                    span("Page ")
                                    currentPageNumber()
                                    span(" of ")
                                    totalPages()
                                }

                            content()
                                .dynamic(OrdersTableWithPageSubtotalsComponent(items))
                        }
                }
            }
            .generatePdf(output("code-pattern-dynamic-component-table-with-per-page-subtotals.pdf"))
    }

    class OrderItem {
        val itemName: String = Placeholders.label()

        // The Kotlin counterpart of C# Placeholders.Random.Next(minValue, maxValue)
        // (System.Random is not bridged): a random integer within [minValue, maxValue).
        val price: Int = Random.nextInt(1, 11) * 10
        val count: Int = Random.nextInt(1, 11)
    }

    class OrdersTableWithPageSubtotalsComponentState {
        var shownItemsCount: Int = 0
    }

    class OrdersTableWithPageSubtotalsComponent(private val items: List<OrderItem>) : IDynamicComponent {
        var state: OrdersTableWithPageSubtotalsComponentState = OrdersTableWithPageSubtotalsComponentState().apply {
            shownItemsCount = 0
        }

        override fun compose(context: DynamicContext): DynamicComponentComposeResult {
            val header = composeHeader(context)
            val sampleFooter = composeFooter(context, emptyList())
            val decorationHeight = header.size.height + sampleFooter.size.height

            val rows = getItemsForPage(context, decorationHeight).toList()
            val footer = composeFooter(context, rows.map { it.first })

            val content = context.createElement {
                shrink().decoration {
                    before().element(header)

                    content().column {
                        for (row in rows)
                            item().element(row.second)
                    }

                    after().element(footer)
                }
            }

            state = OrdersTableWithPageSubtotalsComponentState().apply {
                shownItemsCount = state.shownItemsCount + rows.size
            }

            val result = DynamicComponentComposeResult()
            result.content = content
            result.hasMoreContent = state.shownItemsCount < items.size
            return result
        }

        private fun composeHeader(context: DynamicContext): IDynamicElement {
            return context.createElement {
                width(context.availableSize.width)
                    .borderBottom(1f)
                    .borderColor(Colors.Grey.Darken2)
                    .padding(10f)
                    .defaultTextStyle(TextStyle.Default.semiBold())
                    .row {
                        constantItem(50f).text("#")
                        relativeItem().text("Item name")
                        constantItem(75f).alignRight().text("Count")
                        constantItem(75f).alignRight().text("Price")
                        constantItem(75f).alignRight().text("Total")
                    }
            }
        }

        private fun composeFooter(context: DynamicContext, items: List<OrderItem>): IDynamicElement {
            val total = items.sumOf { it.count * it.price }

            return context.createElement {
                width(context.availableSize.width)
                    .padding(10f)
                    .alignRight()
                    .text("Subtotal: ${total}$")
                    .bold()
            }
        }

        private fun getItemsForPage(context: DynamicContext, decorationHeight: Float): Sequence<Pair<OrderItem, IDynamicElement>> = sequence {
            var totalHeight = decorationHeight

            for (index in state.shownItemsCount until items.size) {
                val item = items.elementAt(index)

                val element = context.createElement {
                    width(context.availableSize.width)
                        .borderBottom(1f)
                        .borderColor(Colors.Grey.Lighten2)
                        .padding(10f)
                        .row {
                            constantItem(50f).text((index + 1).toString())
                            relativeItem().text(item.itemName)
                            constantItem(75f).alignRight().text(item.count.toString())
                            constantItem(75f).alignRight().text("${item.price}$")
                            constantItem(75f).alignRight().text("${item.count * item.price}$")
                        }
                }

                val elementHeight = element.size.height

                // it is important to use the Size.Epsilon constant to avoid floating point comparison issues
                if (totalHeight + elementHeight > context.availableSize.height + Size.Epsilon)
                    break

                totalHeight += elementHeight
                yield(item to element)
            }
        }
    }
}
